--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.c	2025-06-24 23:21:14.810004356 +0800
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.c	2025-06-25 18:58:28.807000462 +0800
@@ -43,6 +43,8 @@
 static int mtk_rest_cnt = 0;
 atomic_t eth1_in_br = ATOMIC_INIT(0);
 EXPORT_SYMBOL(eth1_in_br);
+struct net_device *ppd_dev;
+EXPORT_SYMBOL(ppd_dev);
 static int mtk_msg_level = -1;
 module_param_named(msg_level, mtk_msg_level, int, 0);
 MODULE_PARM_DESC(msg_level, "Message level (-1=defaults,0=none,...,16=all)");
@@ -2445,6 +2447,7 @@
 	int done = 0, bytes = 0;
 	dma_addr_t dma_addr = DMA_MAPPING_ERROR;
 	int ppe_idx = 0;
+	int sent_ppd = 0;
 
 	if (unlikely(!ring))
 		goto rx_done;
@@ -2483,8 +2486,11 @@
 		}
 
 		if ( (mac == 4) || ((FIELD_GET(MTK_RXD4_PPE_CPU_REASON, trxd.rxd4)) == HIT_BIND_FORCE_TO_CPU))
+		{
+		
 		mac = atomic_read(&eth1_in_br);
-
+		sent_ppd = 1;
+		}
 		if (unlikely(mac < 0 || mac >= MTK_MAX_DEVS ||
 			     !eth->netdev[mac]))
 			goto release_desc;
@@ -2643,6 +2649,10 @@
 			hw_lro_flush_stats_update(ring->ring_no, &trxd);
 		}
 
+		if (sent_ppd && ppd_dev){
+		skb->protocol = eth_hdr(skb)->h_proto;
+		skb->dev = ppd_dev;
+		}
 		skb_record_rx_queue(skb, 0);
 		napi_gro_receive(napi, skb);
 		
--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.h	2025-06-24 23:21:14.810004356 +0800
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.h	2025-06-25 18:58:28.807000462 +0800
@@ -1451,11 +1451,11 @@
 #define MT7981_CAPS  (MTK_GMAC1_SGMII | MTK_GMAC2_SGMII | MTK_GMAC2_GEPHY | \
 		      MTK_MUX_GMAC12_TO_GEPHY_SGMII | MTK_QDMA | \
 		      MTK_MUX_U3_GMAC23_TO_QPHY | MTK_U3_COPHY_V2 | \
-		      MTK_RSTCTRL_PPE1 | MTK_SRAM | MTK_PDMA_INT | MTK_HWLRO)
+		      MTK_RSTCTRL_PPE1 | MTK_PDMA_INT)
 
 #define MT7986_CAPS  (MTK_GMAC1_SGMII | MTK_GMAC2_SGMII | \
 		      MTK_MUX_GMAC12_TO_GEPHY_SGMII | MTK_QDMA | \
-		      MTK_RSTCTRL_PPE1 | MTK_SRAM | MTK_PDMA_INT | MTK_HWLRO)
+		      MTK_RSTCTRL_PPE1 | MTK_PDMA_INT)
 
 #define MT7987_CAPS  (MTK_36BIT_DMA | MTK_GMAC1_SGMII | \
 		      MTK_GMAC2_2P5GPHY_V2 | MTK_GMAC2_SGMII | MTK_GMAC3_SGMII | \

 
